src/opi/utils/misc.py

```diff
@@ -173,3 +173,10 @@ def check_minimal_version(version: OrcaVersion, /) -> bool:
     version : OrcaVersion
     """
     return cast(bool, version >= ORCA_MINIMAL_VERSION)
+
+
+T = TypeVar("T")
+
+
+def is_safe_index(seq: Optional[list[T]], i: int) -> TypeGuard[list[T]]:
+    return seq is not None and -len(seq) <= i < len(seq)
```

Comment:
IMO the function should look as follows.
This is the safest to check if an index is valid, as it also makes sure that the object `seq` is subscriptable.
```suggestion
def is_safe_index(seq: Sequence | None, i: int) -> bool:
    """
    ...Docstring..
    """
    try:
        seq[i]
    except (IndexError, TypeError):
        return False
    else:
        return True
```

Side note: `TypeGuard` is wrong here. As it guarantees that the elements of the sequence are of the given type(s), which is not check here at all.